### DISCUSSION TITLE
Unnecessary isinstance call; "Foo" is always an instance of "Foo"

### DISCUSSION THREAD

**Discussion by samvv** [microsoft/pyright#7356] [General]
I'm checking a file in strict mode while doing some refactoring, and was surprised to learn that the following does not typecheck:

```py
class TokenSpec:
    pass

class NodeSpec:
    pass

class VariantSpec:
    pass

type Spec = TokenSpec | NodeSpec | VariantSpec

def foo(spec: Spec) -> str:
    if isinstance(spec, TokenSpec):
        return 'token'
    if isinstance(spec, NodeSpec):
        return 'node'
    if isinstance(spec, VariantSpec):
        return 'variant'
   assert_never(spec)
```

I do this with the `assert_never` so that if there are any new nodes added in the future, I immediately will get a type error.

However, Pyright complains with the following error:

```
Unnecessary isinstance call; "VariantSpec" is always an instance of "VariantSpec"
```

I was wondering: what's the reasoning behind this? The above code seems to me like a perfectly valid pattern to write down, and even the TypeScript compiler does this internally.

**Comment by erictraut** [comment]
It looks like you've enabled the strict-mode `reportUnnecessaryIsininstance` diagnostic check. This check is not enabled by default in standard mode. If you use this check, pyright will check for unnecessary `isinstance` calls — cases where the isinstance call is guaranteed to return false or true every time. If you don't want this diagnostic check, you don't need to enable it.

If you want to keep the check, then you would need to either suppress it in this case or modify your code slightly to avoid the unnecessary isinstance. You can, for example, add an `assert` call and remove the `assert_never` call.

```python
    assert isinstance(spec, VariantSpec):
    return "variant"
```

**Reply by samvv** [reply]
Indeed I enabled strict mode so I got all of the diagnostics.

That last snipped of code is what I needed, thanks!
